pulsar-client-cpp/lib/PublisherStatsImpl.cc

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <lib/PublisherStatsImpl.h>
+#include <lib/LogUtils.h>
+
+namespace pulsar {
+DECLARE_LOG_OBJECT();
+
+    boost::array<double, 4> PublisherStatsImpl::probs = {0.5, 0.9,0.99,0.999};
+
+    std::string PublisherStatsImpl::latencyToString(const LatencyAccumulator& obj) {
+        boost::accumulators::detail::extractor_result<LatencyAccumulator, boost::accumulators::tag::extended_p_square>::type latencies = boost::accumulators::extended_p_square(obj);
+        std::stringstream os;
+        os <<   "Latencies [ 50pct: " << latencies[0] / 1e3 << "ms" <<
+                ", 90pct: " << latencies[1] / 1e3 << "ms" <<
+                ", 99pct: " << latencies[2] / 1e3 << "ms" <<
+                ", 99.9pct: " << latencies[3] / 1e3 << "ms" <<
+                "]";
+        return os.str();
+    }
+
+    double PublisherStatsImpl::differenceInMicros(timespec& start, timespec& end) {
+        return (end.tv_sec - start.tv_sec) * 1e6 + (end.tv_nsec - start.tv_nsec) / 1e3;
+    }
+
+    PublisherStatsImpl::PublisherStatsImpl(std::string producerStr, DeadlineTimerPtr timer, uint64_t statsIntervalInSeconds)
+    : numMsgsSent_(0),
+    numBytesSent_(0),
+    numAcksReceived_(0),
+    totalMsgsSent_(0),
+    totalBytesSent_(0),
+    totalAcksReceived_(0),
+    timer_(timer),
+    producerStr_(producerStr),
+    statsIntervalInSeconds_(statsIntervalInSeconds),
+    mutex_(),
+    latencyAccumulator_(boost::accumulators::tag::extended_p_square::probabilities = probs),
+    totalLatencyAccumulator_(boost::accumulators::tag::extended_p_square::probabilities = probs) {
+        timer_->expires_from_now(boost::posix_time::seconds(statsIntervalInSeconds_));
+        timer_->async_wait(
+                boost::bind(&pulsar::PublisherStatsImpl::flushAndReset, this,
+                        boost::asio::placeholders::error));
+    }
+
+    void PublisherStatsImpl::printStats() {
+        LOG_INFO(*this);
+    }
+
+    void PublisherStatsImpl::flushAndReset(const boost::system::error_code& ec) {
+        if (ec) {
+            LOG_DEBUG("Ignoring timer cancelled event, code[" << ec <<"]");
+            return;
+        }
+        {
+            Lock lock(mutex_);
+            printStats();
```

Comment:
doing I/O with the lock held isn't great.  how about add a swap method to PublisherStatsImpl so that would do this as:

```
// in flushAndReset()
PublisherStatsImpl tmp;
{
    Lock lock(mutex_);
    this->swap();
}
// reset timer here
tmp.printStats();
```

this also shows an interface problem: the base class ```printStats()``` is not safe to call unless you hold ```mutex_``` but there's no way to do that if you have a PublishStatsBase reference.  you could split it into two methods here: printStats that grabs the lock and printStatsUnlocked which doesn't and would be what you use in the snippet above.

why is printStats called both explicitly and on a timer? wouldn't one or the other be the right approach?



```